### PR TITLE
Refactor kernel interface and UFL futzing

### PR DESCRIPTION
**Pull request by miklos1**
This is a refactoring step before we could have support for `CellVolume` and `FacetArea` (see #1).

Major changes:
 * Kernel interface is split off from the driver.
 * UFL routines are moved to `ufl_utils.py` (mostly from `fem.py`)
 * GEM -> ImperoC pipeline moves to `impero_utils.py` (basically from `driver.py`)

Requires firedrakeproject/firedrake#736.

### CODE REVIEW

File: tsfc/kernel_interface.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        self.ast = ast
+        self.integral_type = integral_type
+        self.oriented = oriented
+        self.subdomain_id = subdomain_id
+        self.coefficient_numbers = coefficient_numbers
+        super(Kernel, self).__init__()
+
+
+class Interface(object):
+    def __init__(self, integral_type, subdomain_id):
```
Comment (miklos1) [inline tsfc/kernel_interface.py]: Need some docstrings in this class.

File: tsfc/kernel_interface.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+    def coefficient_mapper(self):
+        return lambda coefficient: self.coefficient_map[coefficient]
+
+    def require_cell_orientations(self):
+        self.kernel.oriented = True
+
+    def construct_kernel(self, name, body):
+        args = [self.local_tensor, self.coordinates_arg]
+        if self.kernel.oriented:
+            args.append(2, coffee.Decl("int *restrict *restrict",
```
Comment (wence-) [inline tsfc/kernel_interface.py]: Don't need the `2, ` I think.

Comment (miklos1) [inline tsfc/kernel_interface.py]: Correct.

File: tsfc/driver.py [hunk 3]
```diff
@@ -183,24 +120,40 @@ def compile_integral(idata, fd, prefix, parameters):
...
+            break
+
+    impero_c = impero_utils.compile_gem([gem.partial_indexed(rv, argument_indices)
+                                         for rv in return_variables], ir,  # TODO
+                                        tuple(quadrature_indices) + argument_indices,
+                                        coffee_licm=parameters["coffee_licm"],
+                                        remove_zeros=True)
+
+    # Generate COFFEE
     index_names = zip(argument_indices, ['j', 'k'])
```
Comment (wence-) [inline tsfc/driver.py]: Could use `index_names = tuple(a.name for a in argument_indices)`

File: tsfc/fem.py [hunk 4]
```diff
@@ -383,18 +145,17 @@ def __getitem__(self, key):
...
         integral_type = tabulation_manager.integral_type
         self.weights = gem.Literal(weights)
         self.quadrature_index = quadrature_index
-        self.argument_indices = argument_indices
         self.tabulation_manager = tabulation_manager
         self.integral_type = integral_type
-        self.coefficient_map = coefficient_map
+        self.coefficient_mapper = coefficient_mapper
         self.index_cache = index_cache
+        self.argument_indices = (gem.Index(name='j'), gem.Index(name='k'))
```
Comment (wence-) [inline tsfc/fem.py]: Is there a reason you *don't* pass argument_indices in any more and instead build new index instances here?

File: tsfc/impero_utils.py [hunk 5]
```diff
@@ -27,6 +28,90 @@
...
+    apply_ordering = make_index_orderer(index_ordering)
+
+    get_indices = lambda expr: apply_ordering(expr.free_indices)
+
+    # Build operation ordering
+    ops = scheduling.emit_operations(zip(return_variables, expressions), get_indices)
+
+    # Empty kernel
+    if len(ops) == 0:
+        raise NoopError()
```
Comment (wence-) [inline tsfc/impero_utils.py]: I don't really like exceptions for control flow, but I don't see a better way of doing this!

**Comment by wence-** [comment]
Mostly looks fine (it's predominantly code movement).

File: tsfc/driver.py (hunk 3 above)
Comment (miklos1) [inline tsfc/driver.py]: Done.

File: tsfc/fem.py (hunk 4 above)
Comment (miklos1) [inline tsfc/fem.py]: I have changed my mind again. I think the current version is less convoluted than both the original and the first proposed.

**Comment by miklos1** [comment]
I have revised this pull request, hitting it to most of the other unmerged branches. I'm now confident that this is a step in the right direction, please have a final review.

File: tsfc/kernel_interface.py [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+        """Construct a COFFEE function declaration with the
+        accumulated glue code.
+
+        :arg name: function name
+        :arg args: function argument list
+        :arg body: function body (:class:`coffee.Block` node)
+        :returns: :class:`coffee.FunDecl` object
+        """
+        body.open_scope = False
+        body_ = coffee.Block(self.prepare + [body] + self.finalise)
```
Comment (wence-) [inline tsfc/kernel_interface.py]: Eww, use `coffee.Block(self.prepare + body.children + self.finalise)`?

Comment (miklos1) [inline tsfc/kernel_interface.py]: Now that I officially assume that `body` is a `coffee.Block`, I guess I can do just that. Maybe I can even add an assertion.

**Comment by wence-** [comment]
OK.

**Comment by miklos1** [comment]
Let's wait for [Travis](https://travis-ci.org/firedrakeproject/firedrake/builds/117716922), then I will merge.

**Pull request merged**